[parquet] Support us precision Time type in batch reader

ZacBlanco opened the pull request:
## Description

Iceberg timestamps are written with microsecond precision as defined by the spec. Before this change, the parquet batch reader optimization did not properly support reading microsecond-precision `time` types. Presto returned results which differed depending on whether the optimization was enabled

This change adds proper support in the batch reader.

## Motivation and Context

Previously Presto would return wrong results in the Iceberg connector when parquet batch read optimizations are enabled.

<details><summary>Behavior before this change:</summary>
<p>


```sql
CREATE TABLE t(i time);
INSERT INTO t VALUES time '1:00';
-- without batch optimization
presto:tpch> SET SESSION iceberg.parquet_batch_read_optimization_enabled = false;
presto:tpch> SELECT CAST(i as bigint) FROM t;
  _col0
----------
 32400000 -- (5 trailing zeroes)
(1 row)


-- with batch optimization
presto:tpch> SET SESSION iceberg.parquet_batch_read_optimization_enabled = true;
SET SESSION
presto:tpch> SELECT CAST(i as bigint) FROM t;
    _col0
-------------
 32400000000 -- (8 trailing zeroes)
(1 row)
```

</p>
</details> 

<details><summary>Behavior after this change:</summary>
<p>


```sql
CREATE TABLE t(i time);
INSERT INTO t VALUES time '1:00';
-- without batch optimization
presto:tpch> SET SESSION iceberg.parquet_batch_read_optimization_enabled = false;
presto:tpch> SELECT CAST(i as bigint) FROM t;
  _col0
----------
 32400000 -- (5 trailing zeroes)
(1 row)


-- with batch optimization
presto:tpch> SET SESSION iceberg.parquet_batch_read_optimization_enabled = true;
SET SESSION
presto:tpch> SELECT CAST(i as bigint) FROM t;
    _col0
-------------
 32400000 -- (5 trailing zeroes)
(1 row)
```

</p>
</details> 




## Impact

Proper `time` type support.

## Test Plan

Added test to verify equivalent return values with and without the optimization in the Iceberg connector.

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== RELEASE NOTES ==

Iceberg Connector Changes
* Fix time-type columns to return properly when ``iceberg.parquet-batch-read-optimization-enabled`` is set to ``TRUE``. :pr:`23542`
```

steveburnett commented:
Nit suggestion for release note entry: 

```
== RELEASE NOTES ==

Iceberg Connector Changes
* Fix time-type columns to return properly when ``iceberg.parquet-batch-read-optimization-enabled`` is set to ``TRUE``. :pr:`23542`
```

hantangwangd commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/Decoders.java:
```diff
@@ -216,6 +223,10 @@ private static ValuesDecoder createValuesDecoder(ColumnDescriptor columnDescript
                         return new Int64TimestampMicrosDeltaBinaryPackedValuesDecoder(valueCount, inputStream);
                     }
 
+                    if (isTimeMicrosType(columnDescriptor)) {
+                        return new Int64TimeMicrosDeltaBinaryPackedValuesDecoder(valueCount, inputStream);
+                    }
+
```
Cannot comment above, but we should also support decoder for TimeType when encoding is `RLE_DICTIONARY` or `PLAIN_DICTIONARY`. Parquet will use `PLAIN_DICTIONARY`(v1) or `RLE_DICTIONARY`(v2) as the encoding if it determines that it's worth using dictionary encoding. For example, if we create a table and insert data as follows:

```
CREATE TABLE time_batch_read(a int, b time);
INSERT INTO time_batch_read VALUES (1, time '1:00:00.000'), (2, time '1:00:00.000'), (3, time '1:00:00.000'), (4, time '1:00:00.000'), (5, time '1:00:00.000'), (6, time '1:00:00.000');
```

Parquet will use dictionary encoding for column `b` on writing the data. In this scenario, we will get an error as currently we haven't supported something like `Int64TimeMicrosRLEDictionaryValuesDecoder`.

ZacBlanco replied:
Good catch! I added support and included additional tests in the `presto-parquet` module and updated the test in the Iceberg connector to verify this scenario

hantangwangd commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java:
```diff
@@ -1829,6 +1829,24 @@ public void testFilterWithRemainingPredicate(boolean pushdownFilterEnabled)
+    @Test
+    public void testBatchReadOnTimeType()
+    {
+        assertQuerySucceeds("CREATE TABLE time_batch_read(i time)");
+        assertUpdate("INSERT INTO time_batch_read VALUES time '1:00:00.000'", 1);
+        Session disabledBatchRead = Session.builder(getSession())
+                .setCatalogSessionProperty("iceberg", PARQUET_BATCH_READ_OPTIMIZATION_ENABLED, "false")
+                .build();
+        Session enabledBatchRead = Session.builder(getSession())
+                .setCatalogSessionProperty("iceberg", PARQUET_BATCH_READ_OPTIMIZATION_ENABLED, "true")
+                .build();
+        @Language("SQL") String query = "SELECT * FROM time_batch_read LIMIT 1";
+        MaterializedResult disabledResult = getQueryRunner().execute(disabledBatchRead, query);
+        MaterializedResult enabledResult = getQueryRunner().execute(enabledBatchRead, query);
+        assertEquals(disabledResult, enabledResult);
+        assertQuerySucceeds("DROP TABLE time_batch_read");
```
We should cover the scenarios for both dictionary encoding and plain encoding. Besides it's better to test for both `v1` and `v2` parquet version.

ZacBlanco replied:
updated

hantangwangd commented on presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/rle/Int64TimeMicrosRLEDictionaryValuesDecoder.java:
```diff
@@ -0,0 +1,113 @@
+
+public class Int64TimeMicrosRLEDictionaryValuesDecoder
+        extends BaseRLEBitPackedDecoder
+        implements ValuesDecoder.Int64TimeMicrosValuesDecoder
```
nit: use static import

ZacBlanco replied:
fixed

hantangwangd reviewed (approved):
Thanks for the fix. Lgtm, one little nit.

yingsu00 reviewed (commented):
@ZacBlanco I see the Time decoders are exactly the same with existing Timestamp ones. There's no point to duplicate so much code. Maybe we can rename the existing Timestamp decoders like Int64TimestampMicrosRLEDictionaryValuesDecoder to Int64TimeAndTimestampMicrosRLEDictionaryValuesDecoder?

ZacBlanco commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java:
```diff
@@ -1829,6 +1832,36 @@ public void testFilterWithRemainingPredicate(boolean pushdownFilterEnabled)
     }
 
+    @DataProvider(name = "parquetVersions")
+    public Object[][] parquetVersionsDataProvider()
```
updated

ZacBlanco replied:
the tests fail if this method is private

ZacBlanco commented:
Thanks for the review @yingsu00 . I agree - let's try not to duplicate code. I updated the naming of the decoders across the module and updated the conditionals to check `isTimestampMicros || isTimeMicros`
